CV: Threshold Analysis (simple verification) (#3340)
harness/developer-hub: docs/continuous-delivery/verify/cv-getstarted/configure-first-cv.md

![Verify step settings page](./static/verify-step-settings-page.png)

## Define name and timeout information

On the Verify settings page, do the following:

1. In **Name**, enter a name for the Verification step.
   
2. In **Timeout**, enter a timeout value for the step. Harness uses this information to time out the verification. Use the following syntax to define a timeout:
   - **w** for weeks.
   - **d** for days.
   - **h** for hours.
   - **m** for minutes.
   - **s** for seconds.
   - **ms** for milliseconds.

   For example, use 1w for one week, 7d for 7 days, 24h for 24 hours, 100m for 100 minutes, 500s for 500 seconds, and 1000ms for 1000 milliseconds.
   
   The maximum timeout value you can set is **53w**. You can also set timeouts at the pipeline level.

## Select the continuous verification type and duration

On the Verify settings page, do the following:

1. In **Continuous Verification Type**, select **Threshold Analysis [No ML]**.
   
2. In **Duration**, choose a duration. Harness uses the data points within this duration for analysis. For instance, if you select 15 minutes, Harness analyzes the first 15 minutes of your health source data.
   

> 3. In the **Artifact Tag** field, reference the primary artifact that you added in the **Artifacts** section of the **Service** tab. Use the Harness expression `<+serviceConfig.artifacts.primary.tag>` to reference the primary artifact. To learn about artifact expression, go to [Harness expression](/docs/platform/Variables-and-Expressions/harness-variables).

Reviewer: ```suggestion
3. In the **Artifact Tag** field, reference the primary artifact that you added in the **Artifacts** section of the **Service** tab. Use the Harness expression `<+serviceConfig.artifacts.primary.tag>` to reference the primary artifact. To learn about artifact expression, go to [Harness expressions](/docs/platform/Variables-and-Expressions/harness-variables).
```

## Create a monitored service

The next step is to create a monitored service for the Verify step. Harness CV uses a monitored service to monitor health trend deviations using logs and metrics obtained from your SumoLogic health source.

:::info note

> When setting up the pipeline, if you've configured service or environment as runtime values, the auto-create option for monitored services won't be available. When you run the pipeline, Harness combines the service and environment values to create a monitored service. If a monitored service with the same name already exists in your project, it will be assigned to the pipeline. If not, Harness skips the verification step.

Reviewer: ```suggestion
When setting up the pipeline, if you've configured a service or an environment as runtime values, the auto-create option for monitored services won't be available. When you run the pipeline, Harness combines the service and environment values to create a monitored service. If a monitored service with the same name already exists in your project, it will be assigned to the pipeline. If not, Harness skips the verification step.
```

#### Define a query

1. In the **Query** box, write a query. As an example, enter the query `metric=memory` for querying the disk usage.
   
2. Select **Run Query**.
   
   Sample data is displayed in the **Records** box. The **Chart** box displays the graph corresponding to the sample data. This helps you verify if the query that you have built is correct.

   ![Query - disk usage](./static/cv-sumologic-select-metric-query-memory.png)

   Disk usage records and chart are being displayed for the query:

   ![Memory usage records and charts](./static/cv-sumologic-select-metric-query-memory-chart-records.png)

### Assign services and configure risk profile

1. In the **Assign** section, select **Continuous Verification (Applied to the pipelines in the Continuous Deployment)**.

2. Under **Risk Category**, select **Performance/Throughput**

3. Under **Deviation Compared To Baseline**, select **Higher counts = higher risk**.

  ![Assign services and configure the risk profile](./static/cv-simple-cv-healthsource-assign-service.png)
  

#### Configure threshold

> Harness CV evaluates the metrics based on the fail-fast threshold you configure. If the metrics breach the fail-fast threshold, the verification step fails. In the context of the [SumoLogic query](#define-a-query) you defined, the purpose here is to set a threshold for memory usage beyond which the application is considered to be in a problematic state, and the verification fails.

Reviewer: ```suggestion
Harness CV evaluates the metrics based on the fail-fast threshold you configure. If the metrics exceed the fail-fast threshold, the verification step fails. In the context of the [SumoLogic query](#define-a-query) you defined, the threshold sets a memory usage limit beyond which the application is considered to be in a problematic state, and the verification fails.
```

> To set fail-fast threshold, follow these steps:

Reviewer: ```suggestion
To set a fail-fast threshold, follow these steps:
```

> 1. Expand **Advanced (Optional)**, go to the **Fail-Fast Thresholds** tab, and then select the **+ Add Threshold** button.

Reviewer: ```suggestion
1. Expand **Advanced (Optional)**, go to the **Fail-Fast Thresholds** tab, and then select **+ Add Threshold**.
```

2. From the **Metric** dropdown, select **SumoLogic Metric**. This is the metric you created in the [Define metric configuration settings](#define-metric-configuration-settings) step.

3. In the **Action** field, select what the CV should do when applying the rule. Let's select **Fail Immediately**.

> 4. In the **Criteria** field, choose **Percentage Deviation**, and from the greater than select **1**. The purpose here is to set a threshold for memory usage beyond which the system or application is considered to be in a problematic state, and action should be taken.

Reviewer: ```suggestion
4. In the **Criteria** field, choose **Percentage Deviation**, and from the greater than select **1**. This sets a threshold for memory usage beyond which the system or application is considered to be in a problematic state, and action should be taken.
```
